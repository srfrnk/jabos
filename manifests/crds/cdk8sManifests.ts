import { Construct } from 'constructs';
import Manifests from './manifests';

export default class Cdk8sManifests extends Manifests {
  constructor(scope: Construct) {
    super(scope, 'Cdk8s', {
      kind: 'Cdk8s',
      singular: 'cdk8s',
      shortNames: ['cdk8s'],
      schema: {
        description: '`Cdk8sManifest` objects define a folder with [Cdk8s](https://cdk8s.io/) based manifests to deploy.',
        properties: {
          spec: {
            required: [],
            properties: {
              commitEnvKey: {
                type: 'string',
                default: 'latestCommitHash',
                description: 'The name of the environment variable to use for injecting the latest git commit hash. The commit hash is used to tag generated images for example. See usage example below.',
              },
              env: {
                type: 'object',
                additionalProperties: true,
                default: {},
                description: 'Key-Values of environment variables to inject when building.',
              },
            }
          },
        }
      }
    });
  }
}
