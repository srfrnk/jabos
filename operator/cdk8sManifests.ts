import { Request, Response, NextFunction } from 'express';

import genericManifests from './genericManifests'
import { Quantity } from './imports/k8s';
import { CustomizeRequest, FinalizeRequest, SyncRequest } from './metaControllerHooks';
import { getRepo } from './misc';

export default {
  async sync(syncRequest: Request, response: Response, next: NextFunction) {
    const request: SyncRequest = syncRequest.body;
    genericManifests.debugRequest('cdk8s', 'sync', request.object, request);

    const repo = getRepo(request);
    const spec: any = request.object.spec;
    const latestCommit = repo.status.latestCommit;

    await genericManifests.sync({
      metricLabel: 'cdk8s',
      type: 'cdk8s',
      metricName: 'cdk8s',
      env: { ...spec.env, [spec.commitEnvKey]: latestCommit },
      volumeMounts: [
        {
          name: 'm2',
          mountPath: "/home/user/.m2"
        },
        {
          name: 'npm',
          mountPath: "/home/user/.npm"
        },
        {
          name: 'local',
          mountPath: "/home/user/.local"
        },
        {
          name: 'go',
          mountPath: "/home/user/go"
        },
        {
          name: 'cache',
          mountPath: "/home/user/.cache"
        }
      ],
      volumes: [
        {
          name: 'm2',
          emptyDir: {}
        },
        {
          name: 'npm',
          emptyDir: {}
        },
        {
          name: 'local',
          emptyDir: {}
        },
        {
          name: 'go',
          emptyDir: {}
        },
        {
          name: 'cache',
          emptyDir: {}
        }
      ],
      resources: {
        limits: {
          memory: Quantity.fromString("2Gi"),
        },
        requests: {
          memory: Quantity.fromString("1Gi"),
        }
      },
      writableGitFolder: true
    }, request, response);
  },

  async customize(customizeRequest: Request, response: Response, next: NextFunction) {
    const request: CustomizeRequest = customizeRequest.body;
    genericManifests.debugRequest('cdk8s', 'customize', request.parent, request);

    await genericManifests.customize('cdk8s', request, response);
  },

  async finalize(finalizeRequest: Request, response: Response, next: NextFunction) {
    const request: FinalizeRequest = finalizeRequest.body;
    genericManifests.debugRequest('cdk8s', 'finalize', request.object, request);

    await genericManifests.finalize('cdk8s', request, response);
  }
}
