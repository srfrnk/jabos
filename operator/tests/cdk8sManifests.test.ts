import { Request, Response } from 'express';
import mockResponse from './mockResponse';
import cdk8sManifests from '../cdk8sManifests';
import { setMock, clearMock } from './settingsMock';

beforeEach(() => {
  setMock();
});

afterEach(() => {
  clearMock();
});

test('cdk8sManifests sync', async () => {
  const req = {
    body: {
      controller: {
        metadata: {
          name: 'controller_name'
        }
      },
      object: {
        kind: 'object_kind',
        metadata: {
          name: 'name_value',
          namespace: 'namespace_value',
          uid: 'uid_value'
        },
        spec: {
          commitEnvKey: 'commitEnvKey_value',
          targetNamespace: 'namespace_value',
          gitRepository: 'gitRepository_value',
          path: '.',
          env: {}
        }
      },
      related: {
        'GitRepository.jabos.io/v1': [
          {
            metadata: {
              namespace: 'namespace_value'
            },
            spec: {
              url: 'url_value',
              branch: 'branch_value'
            },
            status: {
              latestCommit: 'commit_value'
            }
          }
        ]
      },
      attachments: {
        'Job.batch/v1': []
      }
    }
  } as unknown as Request;
  const res = mockResponse();
  await cdk8sManifests.sync(req, res as unknown as Response, () => { /* NOOP */ });
  expect(res.json.mock.calls[0]).toMatchSnapshot();
  expect(res.status).toHaveBeenCalledWith(200);
});

test('cdk8sManifests customize', async () => {
  const req = {
    body: {
      parent: {
        metadata: {
          name: 'name_value',
          namespace: 'namespace_value'
        },
        spec: {
          gitRepository: 'gitRepository_value'
        }
      },
    }
  } as unknown as Request;
  const res = mockResponse();
  await cdk8sManifests.customize(req, res as unknown as Response, () => { /* NOOP */ });
  expect(res.json.mock.calls[0]).toMatchSnapshot();
  expect(res.status).toHaveBeenCalledWith(200);
});

test('cdk8sManifests finalize leave', async () => {
  const req = {
    body: {
      controller: {
        metadata: {
          name: 'controller_name'
        }
      },
      object: {
        metadata: {
          name: 'name_value',
          namespace: 'namespace_value',
          annotations: {}
        },
        spec: {
          cleanupPolicy: 'Leave'
        }
      },
      attachments: {
        'Job.batch/v1': []
      }
    }
  } as unknown as Request;
  const res = mockResponse();
  await cdk8sManifests.finalize(req, res as unknown as Response, () => { /* NOOP */ });
  expect(res.json.mock.calls[0]).toMatchSnapshot();
  expect(res.status).toHaveBeenCalledWith(200);
});

test('cdk8sManifests finalize delete', async () => {
  const req = {
    body: {
      controller: {
        metadata: {
          name: 'controller_name'
        }
      },
      object: {
        kind: 'object_kind',
        metadata: {
          name: 'name_value',
          namespace: 'namespace_value',
          annotations: { deployedManifest: 'deployedManifest_value' },
          uid: 'object_uid'
        },
        spec: {
          cleanupPolicy: 'Delete',
          targetNamespace: 'namespace_value',
        }
      },
      attachments: {
        'Job.batch/v1': []
      }
    }
  } as unknown as Request;
  const res = mockResponse();
  await cdk8sManifests.finalize(req, res as unknown as Response, () => { /* NOOP */ });
  expect(res.json.mock.calls[0]).toMatchSnapshot();
  expect(res.status).toHaveBeenCalledWith(200);
});