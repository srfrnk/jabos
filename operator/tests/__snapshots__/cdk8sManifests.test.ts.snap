// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cdk8sManifests customize 1`] = `
Array [
  Object {
    "relatedResources": Array [
      Object {
        "apiVersion": "jabos.io/v1",
        "names": Array [
          "gitRepository_value",
        ],
        "namespace": "namespace_value",
        "resource": "git-repositories",
      },
    ],
  },
]
`;

exports[`cdk8sManifests finalize delete 1`] = `
Array [
  Object {
    "annotations": Object {},
    "attachments": Array [
      Object {
        "apiVersion": "rbac.authorization.k8s.io/v1",
        "kind": "Role",
        "metadata": Object {
          "name": "cleaner-name_value",
          "namespace": "namespace_value",
        },
        "rules": Array [
          Object {
            "apiGroups": Array [
              "*",
            ],
            "resources": Array [
              "*",
            ],
            "verbs": Array [
              "delete",
            ],
          },
        ],
      },
      Object {
        "apiVersion": "v1",
        "kind": "ServiceAccount",
        "metadata": Object {
          "name": "cleaner-name_value",
          "namespace": "namespace_value",
        },
      },
      Object {
        "apiVersion": "rbac.authorization.k8s.io/v1",
        "kind": "RoleBinding",
        "metadata": Object {
          "name": "cleaner-name_value",
          "namespace": "namespace_value",
        },
        "roleRef": Object {
          "apiGroup": "rbac.authorization.k8s.io",
          "kind": "Role",
          "name": "cleaner-name_value",
        },
        "subjects": Array [
          Object {
            "kind": "ServiceAccount",
            "name": "cleaner-name_value",
            "namespace": "namespace_value",
          },
        ],
      },
      Object {
        "apiVersion": "batch/v1",
        "kind": "Job",
        "metadata": Object {
          "labels": Object {
            "type": "manifest-cleaner",
          },
          "name": "manifest-clean-name_value",
        },
        "spec": Object {
          "activeDeadlineSeconds": 1234,
          "backoffLimit": 1234,
          "completionMode": "NonIndexed",
          "completions": 1,
          "parallelism": 1,
          "template": Object {
            "metadata": Object {
              "annotations": Object {
                "manifests": "deployedManifest_value",
              },
              "labels": Object {
                "builder": "manifest-clean-name_value",
              },
              "name": "manifest-clean-name_value",
            },
            "spec": Object {
              "containers": Array [
                Object {
                  "args": Array [],
                  "env": Array [
                    Object {
                      "name": "NAMESPACE",
                      "value": "namespace_value",
                    },
                    Object {
                      "name": "NAME",
                      "value": "name_value",
                    },
                    Object {
                      "name": "OBJECT_UID",
                      "value": "object_uid",
                    },
                    Object {
                      "name": "KIND",
                      "value": "object_kind",
                    },
                    Object {
                      "name": "CONTROLLER",
                      "value": "controller_name",
                    },
                    Object {
                      "name": "JABOS_OPERATOR_URL",
                      "value": "http://operator.settings_namespace:1234/",
                    },
                  ],
                  "image": "settings_imagePrefixmanifest-cleaner:settings_build_number",
                  "imagePullPolicy": "Always",
                  "name": "manifest-cleaner",
                  "resources": Object {
                    "limits": Object {
                      "cpu": "500m",
                      "memory": "500Mi",
                    },
                    "requests": Object {
                      "cpu": "100m",
                      "memory": "100Mi",
                    },
                  },
                  "securityContext": Object {
                    "allowPrivilegeEscalation": false,
                    "capabilities": Object {
                      "drop": Array [
                        "ALL",
                      ],
                    },
                    "readOnlyRootFilesystem": true,
                    "runAsNonRoot": true,
                  },
                  "stdin": true,
                  "tty": true,
                  "volumeMounts": Array [
                    Object {
                      "mountPath": "/manifests",
                      "name": "manifests",
                    },
                    Object {
                      "mountPath": "/tmp",
                      "name": "temp",
                    },
                  ],
                },
              ],
              "restartPolicy": "Never",
              "securityContext": Object {
                "runAsNonRoot": true,
              },
              "serviceAccountName": "cleaner-name_value",
              "volumes": Array [
                Object {
                  "emptyDir": Object {},
                  "name": "temp",
                },
                Object {
                  "downwardAPI": Object {
                    "items": Array [
                      Object {
                        "fieldRef": Object {
                          "fieldPath": "metadata.annotations['manifests']",
                        },
                        "path": "manifests.tar.gz.b64",
                      },
                    ],
                  },
                  "name": "manifests",
                },
              ],
            },
          },
          "ttlSecondsAfterFinished": 1234,
        },
      },
    ],
    "finalized": false,
  },
]
`;

exports[`cdk8sManifests finalize leave 1`] = `
Array [
  Object {
    "annotations": Object {},
    "attachments": Array [],
    "finalized": true,
  },
]
`;

exports[`cdk8sManifests sync 1`] = `
Array [
  Object {
    "attachments": Array [
      Object {
        "apiVersion": "rbac.authorization.k8s.io/v1",
        "kind": "Role",
        "metadata": Object {
          "name": "deployer-name_value",
          "namespace": "namespace_value",
        },
        "rules": Array [
          Object {
            "apiGroups": Array [
              "*",
            ],
            "resources": Array [
              "*",
            ],
            "verbs": Array [
              "get",
              "list",
              "watch",
              "create",
              "update",
              "patch",
            ],
          },
        ],
      },
      Object {
        "apiVersion": "rbac.authorization.k8s.io/v1",
        "kind": "RoleBinding",
        "metadata": Object {
          "name": "deployer-name_value",
          "namespace": "namespace_value",
        },
        "roleRef": Object {
          "apiGroup": "rbac.authorization.k8s.io",
          "kind": "Role",
          "name": "deployer-name_value",
        },
        "subjects": Array [
          Object {
            "kind": "ServiceAccount",
            "name": "builder-gitRepository_value",
            "namespace": "namespace_value",
          },
        ],
      },
      Object {
        "apiVersion": "batch/v1",
        "kind": "Job",
        "metadata": Object {
          "labels": Object {
            "manifest-builder-type": "cdk8s-manifests",
            "type": "manifest-builder",
          },
          "name": "manifest-name_value-commit_value",
        },
        "spec": Object {
          "activeDeadlineSeconds": 1234,
          "backoffLimit": 1234,
          "completionMode": "NonIndexed",
          "completions": 1,
          "parallelism": 1,
          "template": Object {
            "metadata": Object {
              "labels": Object {
                "builder": "manifest-name_value-commit_value",
              },
              "name": "manifest-name_value-commit_value",
            },
            "spec": Object {
              "containers": Array [
                Object {
                  "args": Array [],
                  "env": Array [
                    Object {
                      "name": "TYPE",
                      "value": "cdk8s-manifests",
                    },
                    Object {
                      "name": "NAME",
                      "value": "name_value",
                    },
                    Object {
                      "name": "COMMIT",
                      "value": "commit_value",
                    },
                    Object {
                      "name": "METRIC_NAME",
                      "value": "cdk8sManifestsBuilder",
                    },
                    Object {
                      "name": "METRIC_LABELS",
                      "value": "{\\"namespace\\":\\"namespace_value\\",\\"cdk8s_manifests\\":\\"name_value\\"}",
                    },
                    Object {
                      "name": "NAMESPACE",
                      "value": "namespace_value",
                    },
                    Object {
                      "name": "OBJECT_UID",
                      "value": "uid_value",
                    },
                    Object {
                      "name": "JABOS_OPERATOR_URL",
                      "value": "http://operator.settings_namespace:1234/",
                    },
                  ],
                  "image": "settings_imagePrefixpost-builder:settings_build_number",
                  "imagePullPolicy": "Always",
                  "name": "post-builder",
                  "resources": Object {
                    "limits": Object {
                      "cpu": "500m",
                      "memory": "500Mi",
                    },
                    "requests": Object {
                      "cpu": "100m",
                      "memory": "100Mi",
                    },
                  },
                  "securityContext": Object {
                    "allowPrivilegeEscalation": false,
                    "capabilities": Object {
                      "drop": Array [
                        "ALL",
                      ],
                    },
                    "readOnlyRootFilesystem": true,
                    "runAsNonRoot": true,
                  },
                  "volumeMounts": Array [
                    Object {
                      "mountPath": "/timer",
                      "name": "timer",
                      "readOnly": true,
                    },
                    Object {
                      "mountPath": "/tmp",
                      "name": "temp",
                    },
                  ],
                },
              ],
              "initContainers": Array [
                Object {
                  "args": Array [],
                  "env": Array [
                    Object {
                      "name": "URL",
                      "value": "url_value",
                    },
                    Object {
                      "name": "BRANCH",
                      "value": "branch_value",
                    },
                    Object {
                      "name": "COMMIT",
                      "value": "commit_value",
                    },
                    Object {
                      "name": "METRIC_NAME",
                      "value": "cdk8sManifestsBuilder",
                    },
                    Object {
                      "name": "METRIC_LABELS",
                      "value": "{\\"namespace\\":\\"namespace_value\\",\\"cdk8s_manifests\\":\\"name_value\\"}",
                    },
                    Object {
                      "name": "NAMESPACE",
                      "value": "namespace_value",
                    },
                    Object {
                      "name": "NAME",
                      "value": "name_value",
                    },
                    Object {
                      "name": "OBJECT_UID",
                      "value": "uid_value",
                    },
                    Object {
                      "name": "JABOS_OPERATOR_URL",
                      "value": "http://operator.settings_namespace:1234/",
                    },
                  ],
                  "image": "settings_imagePrefixpre-builder:settings_build_number",
                  "imagePullPolicy": "Always",
                  "name": "pre-builder",
                  "resources": Object {
                    "limits": Object {
                      "cpu": "500m",
                      "memory": "500Mi",
                    },
                    "requests": Object {
                      "cpu": "100m",
                      "memory": "100Mi",
                    },
                  },
                  "securityContext": Object {
                    "allowPrivilegeEscalation": false,
                    "capabilities": Object {
                      "drop": Array [
                        "ALL",
                      ],
                    },
                    "readOnlyRootFilesystem": true,
                    "runAsNonRoot": true,
                  },
                  "volumeMounts": Array [
                    Object {
                      "mountPath": "/gitTemp",
                      "name": "git-temp",
                    },
                    Object {
                      "mountPath": "/tmp",
                      "name": "temp",
                    },
                    Object {
                      "mountPath": "/timer",
                      "name": "timer",
                    },
                  ],
                },
                Object {
                  "args": Array [],
                  "env": Array [
                    Object {
                      "name": "SRC_PATH",
                      "value": ".",
                    },
                    Object {
                      "name": "KIND",
                      "value": "object_kind",
                    },
                    Object {
                      "name": "CONTROLLER",
                      "value": "controller_name",
                    },
                    Object {
                      "name": "commitEnvKey_value",
                      "value": "commit_value",
                    },
                    Object {
                      "name": "NAMESPACE",
                      "value": "namespace_value",
                    },
                    Object {
                      "name": "NAME",
                      "value": "name_value",
                    },
                    Object {
                      "name": "OBJECT_UID",
                      "value": "uid_value",
                    },
                  ],
                  "image": "settings_imagePrefixcdk8s-manifest-builder:settings_build_number",
                  "imagePullPolicy": "Always",
                  "name": "cdk8s-manifest-builder",
                  "resources": Object {
                    "limits": Object {
                      "cpu": "500m",
                      "memory": "2Gi",
                    },
                    "requests": Object {
                      "cpu": "100m",
                      "memory": "1Gi",
                    },
                  },
                  "securityContext": Object {
                    "allowPrivilegeEscalation": false,
                    "capabilities": Object {
                      "drop": Array [
                        "ALL",
                      ],
                    },
                    "readOnlyRootFilesystem": true,
                    "runAsNonRoot": true,
                  },
                  "stdin": true,
                  "tty": true,
                  "volumeMounts": Array [
                    Object {
                      "mountPath": "/home/user/.m2",
                      "name": "m2",
                    },
                    Object {
                      "mountPath": "/home/user/.npm",
                      "name": "npm",
                    },
                    Object {
                      "mountPath": "/home/user/.local",
                      "name": "local",
                    },
                    Object {
                      "mountPath": "/home/user/go",
                      "name": "go",
                    },
                    Object {
                      "mountPath": "/home/user/.cache",
                      "name": "cache",
                    },
                    Object {
                      "mountPath": "/gitTemp",
                      "name": "git-temp",
                    },
                    Object {
                      "mountPath": "/manifests",
                      "name": "manifests",
                    },
                    Object {
                      "mountPath": "/tmp",
                      "name": "temp",
                    },
                    Object {
                      "mountPath": "/build",
                      "name": "build",
                    },
                  ],
                },
                Object {
                  "args": Array [
                    "namespace_value",
                    "cdk8s-manifests",
                    "name_value",
                  ],
                  "env": Array [
                    Object {
                      "name": "NAMESPACE",
                      "value": "namespace_value",
                    },
                    Object {
                      "name": "TARGET_NAMESPACE",
                      "value": "namespace_value",
                    },
                    Object {
                      "name": "TYPE",
                      "value": "cdk8s-manifests",
                    },
                    Object {
                      "name": "NAME",
                      "value": "name_value",
                    },
                    Object {
                      "name": "KIND",
                      "value": "object_kind",
                    },
                    Object {
                      "name": "CONTROLLER",
                      "value": "controller_name",
                    },
                    Object {
                      "name": "NAMESPACE",
                      "value": "namespace_value",
                    },
                    Object {
                      "name": "NAME",
                      "value": "name_value",
                    },
                    Object {
                      "name": "OBJECT_UID",
                      "value": "uid_value",
                    },
                  ],
                  "image": "settings_imagePrefixmanifest-deployer:settings_build_number",
                  "imagePullPolicy": "Always",
                  "name": "manifest-deployer",
                  "resources": Object {
                    "limits": Object {
                      "cpu": "500m",
                      "memory": "500Mi",
                    },
                    "requests": Object {
                      "cpu": "100m",
                      "memory": "100Mi",
                    },
                  },
                  "securityContext": Object {
                    "allowPrivilegeEscalation": false,
                    "capabilities": Object {
                      "drop": Array [
                        "ALL",
                      ],
                    },
                    "readOnlyRootFilesystem": true,
                    "runAsNonRoot": true,
                  },
                  "stdin": true,
                  "tty": true,
                  "volumeMounts": Array [
                    Object {
                      "mountPath": "/manifests",
                      "name": "manifests",
                      "readOnly": true,
                    },
                    Object {
                      "mountPath": "/tmp",
                      "name": "temp",
                    },
                    Object {
                      "mountPath": "/build",
                      "name": "build",
                    },
                  ],
                },
              ],
              "restartPolicy": "Never",
              "securityContext": Object {
                "runAsNonRoot": true,
              },
              "serviceAccountName": "builder-gitRepository_value",
              "volumes": Array [
                Object {
                  "emptyDir": Object {},
                  "name": "m2",
                },
                Object {
                  "emptyDir": Object {},
                  "name": "npm",
                },
                Object {
                  "emptyDir": Object {},
                  "name": "local",
                },
                Object {
                  "emptyDir": Object {},
                  "name": "go",
                },
                Object {
                  "emptyDir": Object {},
                  "name": "cache",
                },
                Object {
                  "emptyDir": Object {},
                  "name": "manifests",
                },
                Object {
                  "emptyDir": Object {},
                  "name": "git-temp",
                },
                Object {
                  "emptyDir": Object {},
                  "name": "temp",
                },
                Object {
                  "emptyDir": Object {},
                  "name": "build",
                },
                Object {
                  "emptyDir": Object {},
                  "name": "timer",
                },
              ],
            },
          },
          "ttlSecondsAfterFinished": 1234,
        },
      },
    ],
    "status": Object {
      "conditions": Array [
        Object {
          "status": "False",
          "type": "Synced",
        },
      ],
    },
  },
]
`;
